File: corehydror/tests/testthat/test-data.R
# Behavioural tests for the data layer: analysis_data(), its summary, and the threshold
# diagnostics. Numeric oracles live in fixtures/ (threshold_diagnostics.json and the censored
# estimation fixtures); what is checked here is the R surface -- spec assembly, argument
# validation, and the invariants a caller relies on.

peaks <- c(12500, 15300, 8900, 22100, 18700, 14200, 9800, 28500, 17400, 11600,
           19200, 13800, 25600, 10500, 16900)

test_that("analysis_data assembles each series and indexes them from zero", {
  d <- analysis_data(
    exact = peaks,
    interval = data.frame(lower = 30000, value = 35000, upper = 40000, index = 15),
    threshold = data.frame(start_index = 15, end_index = 54, value = 28000, number_above = 1),
    uncertain = list(distribution("Normal", c(14000, 2000)))
  )
  expect_s3_class(d, "corehydro_data")
  expect_length(d$exact, length(peaks))
  expect_equal(d$exact[[1]]$index, 0L)
  expect_equal(d$exact[[length(peaks)]]$index, length(peaks) - 1L)
  expect_equal(d$exact[[1]]$value, peaks[1])
  expect_equal(d$interval[[1]]$index, 15L)
  expect_equal(d$threshold[[1]]$number_above, 1L)
  expect_equal(d$uncertain[[1]]$distribution$family, "Normal")
})

test_that("analysis_data accepts a plain vector or a data frame for the exact series", {
  a <- analysis_data(exact = peaks)
  b <- analysis_data(exact = data.frame(value = peaks))
  expect_identical(a, b)
})

test_that("analysis_data rejects malformed input", {
  expect_error(analysis_data(), "at least one of")
  expect_error(analysis_data(interval = data.frame(lower = 1, upper = 3)), "`value` column")
  expect_error(
    analysis_data(interval = data.frame(lower = 5, value = 1, upper = 9)),
    "lower <= value <= upper"
  )
  expect_error(
    analysis_data(threshold = data.frame(start_index = 5, end_index = 1, value = 1,
                                         number_above = 0)),
    "end_index >= start_index"
  )
  expect_error(analysis_data(exact = peaks, index = 1:3), "unused argument")
  expect_error(
    analysis_data(exact = data.frame(value = peaks, index = seq_along(peaks) - 2L)),
    "non-negative"
  )
  expect_error(analysis_data(exact = peaks, uncertain = list(1, 2)), "corehydro_dist")
  expect_error(
    analysis_data(peaks, low_outlier_threshold = 100, mgbt_low_outliers = TRUE),
    "do not also supply"
  )
})

test_that("the summary reports the record diagnostics", {
  s <- analysis_data_summary(analysis_data(peaks))
  expect_equal(s$value, peaks)
  expect_equal(s$index, seq_along(peaks) - 1L)
  expect_equal(s$exact_count, length(peaks))
  expect_equal(s$total_record_length, length(peaks))
  expect_equal(s$number_of_low_outliers, 0L)
  expect_true(all(s$plotting_position > 0 & s$plotting_position < 1))
  # Weibull positions on an uncensored record are rank / (n + 1).
  expect_equal(sort(s$plotting_position), seq_len(length(peaks)) / (length(peaks) + 1))
})

test_that("a plain vector and an equivalent frame summarize identically", {
  expect_identical(analysis_data_summary(peaks), analysis_data_summary(analysis_data(peaks)))
})

test_that("censored series lengthen the record and shift the plotting positions", {
  plain <- analysis_data_summary(analysis_data(peaks))
  censored <- analysis_data_summary(analysis_data(
    exact = peaks,
    threshold = data.frame(start_index = 15, end_index = 54, value = 28000, number_above = 1)
  ))
  expect_equal(censored$total_record_length, 55L)
  expect_equal(censored$threshold_count, 1L)
  expect_false(isTRUE(all.equal(plain$plotting_position, censored$plotting_position)))
})

test_that("the plotting parameter selects the convention and is range-checked", {
  weibull <- analysis_data_summary(analysis_data(peaks))
  cunnane <- analysis_data_summary(analysis_data(peaks), plotting_parameter = 0.40)
  expect_equal(weibull$plotting_parameter, 0)
  expect_equal(cunnane$plotting_parameter, 0.40)
  expect_false(isTRUE(all.equal(weibull$plotting_position, cunnane$plotting_position)))
  expect_error(analysis_data_summary(analysis_data(peaks), plotting_parameter = 1), "\\[0, 1\\)")
  expect_error(analysis_data_summary(analysis_data(peaks), plotting_parameter = -0.1), "\\[0, 1\\)")
})

test_that("MGBT censoring flags the low floods and matches an explicit threshold", {
  low <- c(peaks, 120, 95)
  expect_equal(mgbt_test(low), 2L)

  s <- analysis_data_summary(analysis_data(low, mgbt_low_outliers = TRUE))
  expect_equal(s$number_of_low_outliers, 2L)
  expect_equal(s$low_outlier_threshold, 8900)
  expect_equal(sum(s$is_low_outlier), 2L)
  expect_true(all(s$value[s$is_low_outlier] < 8900))

  # An explicit threshold at the value MGBT chose must censor the same points.
  explicit <- analysis_data_summary(analysis_data(low, low_outlier_threshold = 8900))
  expect_equal(explicit$number_of_low_outliers, 2L)
  expect_identical(explicit$is_low_outlier, s$is_low_outlier)
})

test_that("a threshold censoring more than half the record is rejected", {
  expect_error(
    analysis_data_summary(analysis_data(peaks, low_outlier_threshold = 20000)),
    "50 percent"
  )
})

test_that("threshold_diagnostics returns parallel vectors for each method", {
  x <- dist_random(distribution("GeneralizedPareto", c(0, 100, 0.1)), 300, seed = 4242)

  mrl <- threshold_diagnostics(x, u_min = 0, u_max = 200, n_thresholds = 8)
  expect_named(mrl, c("threshold", "exceedance_count", "mean_excess", "lower_ci", "upper_ci",
                      "modified_scale", "modified_scale_lower_ci", "modified_scale_upper_ci",
                      "shape", "shape_lower_ci", "shape_upper_ci"))
  n <- length(mrl$threshold)
  expect_true(n > 0 && n <= 8)
  expect_length(mrl$mean_excess, n)
  expect_length(mrl$exceedance_count, n)
  expect_length(mrl$modified_scale, 0)
  expect_true(all(mrl$lower_ci < mrl$mean_excess))
  expect_true(all(mrl$upper_ci > mrl$mean_excess))
  expect_true(all(diff(mrl$exceedance_count) <= 0))

  st <- threshold_diagnostics(x, u_min = 0, u_max = 200, n_thresholds = 8,
                              method = "parameter_stability")
  m <- length(st$threshold)
  expect_length(st$shape, m)
  expect_length(st$mean_excess, 0)
  expect_true(all(st$shape_lower_ci < st$shape))
  # The sample is GPD(0, 100, 0.1), so the fitted shape should sit near 0.1.
  expect_lt(abs(st$shape[1] - 0.1), 0.15)
})

test_that("threshold_diagnostics validates its method argument", {
  expect_error(threshold_diagnostics(peaks, 0, 1, method = "nope"), "'arg' should be one of")
})

test_that("print shows the series composition", {
  d <- analysis_data(exact = peaks, mgbt_low_outliers = TRUE)
  expect_output(print(d), "15 exact")
  expect_output(print(d), "Multiple Grubbs-Beck")
  expect_output(print(analysis_data(peaks, low_outlier_threshold = 9000)), "9000")
})

# Harricana River annual peaks (Bobee & Ashkar 1991, Table 1.2, page 5); the oracle for
# mann_kendall (0.7757) is ExactDataHypothesisTests.Test_MannKendall, also transcribed into
# fixtures/data/data_frame_facades.json and core/tests/test_data_frame_facades.cpp.
harricana <- c(
  122, 244, 214, 173, 229, 156, 212, 263, 146, 183, 161, 205, 135, 331, 225,
  174, 98.8, 149, 238, 262, 132, 235, 216, 240, 230, 192, 195, 172, 173, 172,
  153, 142, 317, 161, 201, 204, 194, 164, 183, 161, 167, 179, 185, 117, 192,
  337, 125, 166, 99.1, 202, 230, 158, 262, 154, 164, 182, 164, 183, 171, 250,
  184, 205, 237, 177, 239, 187, 180, 173, 174
)

test_that("analysis_data_hypothesis_test reproduces the Harricana Mann-Kendall oracle", {
  p <- analysis_data_hypothesis_test(harricana, "mann_kendall")
  expect_named(p, "mann_kendall")
  expect_equal(unname(p), 0.7757, tolerance = 1e-3)
})

test_that("analysis_data_hypothesis_test accepts a corehydro_data frame (exact series only)", {
  p_vec <- analysis_data_hypothesis_test(harricana, "mann_kendall")
  p_obj <- analysis_data_hypothesis_test(analysis_data(harricana), "mann_kendall")
  expect_equal(p_vec, p_obj)
})

test_that("analysis_data_hypothesis_test runs a two-sample test given an index", {
  p <- analysis_data_hypothesis_test(harricana, "mann_whitney", index = 50)
  expect_equal(unname(p), 0.5892, tolerance = 1e-2)
})

test_that("analysis_data_hypothesis_test validates method and the two-sample index", {
  expect_error(analysis_data_hypothesis_test(harricana, "not_a_method"), "'arg' should be one of")
  expect_error(analysis_data_hypothesis_test(harricana, "equal_variance_t"), "requires .index.")
})

test_that("analysis_data_statistics returns the twenty named summary statistics", {
  s <- analysis_data_statistics(peaks)
  expect_length(s$value, 20)
  expect_named(s$value, c(
    "Record Length", "Events Per Index (λ)", "Low Outliers", "Minimum", "Maximum", "Mean",
    "Std Dev", "Skewness", "Kurtosis", "Mean (of log)", "Std Dev (of log)", "Skewness (of log)",
    "Kurtosis (of log)", "1%", "5%", "25%", "50%", "75%", "95%", "99%"
  ))
  expect_true(is.finite(s$value[["Mean"]]))  # peaks has 15 >= 10 exact points
})

test_that("analysis_data_statistics(all_data = TRUE) uses the plotting-position fit", {
  exact_only <- analysis_data_statistics(harricana)
  all_data <- analysis_data_statistics(harricana, all_data = TRUE)
  expect_equal(unname(exact_only$value[["Record Length"]]), 69)
  expect_equal(unname(exact_only$value[["Minimum"]]), 98.8)
  expect_equal(unname(all_data$value[["Minimum"]]), 98.8)
  # The two methods use different estimators, so their means need not agree even with no
  # censored data in the frame.
  expect_false(isTRUE(all.equal(
    unname(exact_only$value[["Mean"]]), unname(all_data$value[["Mean"]])
  )))
})

test_that("analysis_data_statistics(standardized = TRUE) adds parallel standardized columns", {
  s <- analysis_data_statistics(harricana, standardized = TRUE)
  expect_length(s$standardized_value, length(harricana))
  expect_length(s$standardized_log10_value, length(harricana))
  expect_true(all(is.finite(s$standardized_value)))
  expect_true(all(is.finite(s$standardized_log10_value)))
})

test_that("analysis_data_statistics reports NaN for a short record", {
  short <- harricana[1:9]
  s <- analysis_data_statistics(short)
  expect_true(all(is.na(s$value)))
})

File: corehydropy/tests/test_data.py
"""Behavioural tests for the data layer: AnalysisData, its summary, and threshold diagnostics.

Numeric oracles live in ``fixtures/`` (threshold_diagnostics.json and the censored estimation
fixtures); what is checked here is the Python surface -- spec assembly, argument validation, the
invariants a caller relies on, and agreement with the R package's identical surface.
"""

from __future__ import annotations

import math

import pytest

import corehydropy as ch

PEAKS = [12500.0, 15300.0, 8900.0, 22100.0, 18700.0, 14200.0, 9800.0, 28500.0, 17400.0,
         11600.0, 19200.0, 13800.0, 25600.0, 10500.0, 16900.0]


def test_analysis_data_assembles_each_series_indexed_from_zero():
    d = ch.analysis_data(
        exact=PEAKS,
        interval={"index": [15], "lower": [30000.0], "value": [35000.0], "upper": [40000.0]},
        threshold={"start_index": 15, "end_index": 54, "value": 28000.0, "number_above": 1},
        uncertain=[ch.Distribution("Normal", [14000, 2000])],
    )
    assert isinstance(d, ch.AnalysisData)
    assert len(d.spec["exact"]) == len(PEAKS)
    assert d.spec["exact"][0]["index"] == 0
    assert d.spec["exact"][-1]["index"] == len(PEAKS) - 1
    assert d.spec["exact"][0]["value"] == PEAKS[0]
    assert d.spec["interval"][0]["index"] == 15
    assert d.spec["threshold"][0]["number_above"] == 1
    assert d.spec["uncertain"][0]["distribution"]["family"] == "Normal"


def test_analysis_data_accepts_a_sequence_or_a_mapping_for_exact():
    assert ch.analysis_data(exact=PEAKS).spec == ch.analysis_data(exact={"value": PEAKS}).spec


@pytest.mark.parametrize(
    "kwargs,message",
    [
        ({}, "at least one of"),
        ({"interval": {"lower": [1], "upper": [3]}}, "`value` column"),
        ({"interval": {"lower": [5], "value": [1], "upper": [9]}}, "lower <= value <= upper"),
        (
            {"threshold": {"start_index": 5, "end_index": 1, "value": 1, "number_above": 0}},
            "end_index >= start_index",
        ),
        ({"exact": {"value": PEAKS, "index": [-1] * len(PEAKS)}}, "non-negative"),
        ({"exact": PEAKS, "low_outlier_threshold": 100, "mgbt_low_outliers": True},
         "do not also supply"),
    ],
)
def test_analysis_data_rejects_malformed_input(kwargs, message):
    with pytest.raises((ValueError, TypeError), match=message):
        ch.analysis_data(**kwargs)


def test_analysis_data_rejects_a_non_distribution_uncertain_series():
    with pytest.raises(TypeError, match="Distribution"):
        ch.analysis_data(exact=PEAKS, uncertain=[1, 2])


def test_summary_reports_the_record_diagnostics():
    s = ch.analysis_data_summary(ch.analysis_data(PEAKS))
    assert s["value"] == PEAKS
    assert s["index"] == list(range(len(PEAKS)))
    assert s["exact_count"] == len(PEAKS)
    assert s["total_record_length"] == len(PEAKS)
    assert s["number_of_low_outliers"] == 0
    assert all(0 < p < 1 for p in s["plotting_position"])
    # Weibull positions on an uncensored record are rank / (n + 1).
    n = len(PEAKS)
    assert sorted(s["plotting_position"]) == pytest.approx([(i + 1) / (n + 1) for i in range(n)])


def test_a_plain_sequence_and_an_equivalent_frame_summarize_identically():
    assert ch.analysis_data_summary(PEAKS) == ch.analysis_data_summary(ch.analysis_data(PEAKS))


def test_censored_series_lengthen_the_record_and_shift_the_positions():
    plain = ch.analysis_data_summary(ch.analysis_data(PEAKS))
    censored = ch.analysis_data_summary(ch.analysis_data(
        exact=PEAKS,
        threshold={"start_index": 15, "end_index": 54, "value": 28000.0, "number_above": 1},
    ))
    assert censored["total_record_length"] == 55
    assert censored["threshold_count"] == 1
    assert censored["plotting_position"] != plain["plotting_position"]


def test_plotting_parameter_selects_the_convention_and_is_range_checked():
    weibull = ch.analysis_data_summary(ch.analysis_data(PEAKS))
    cunnane = ch.analysis_data_summary(ch.analysis_data(PEAKS), plotting_parameter=0.40)
    assert weibull["plotting_parameter"] == 0.0
    assert cunnane["plotting_parameter"] == 0.40
    assert weibull["plotting_position"] != cunnane["plotting_position"]
    for bad in (1.0, -0.1, math.nan):
        with pytest.raises(ValueError, match=r"\[0, 1\)"):
            ch.analysis_data_summary(ch.analysis_data(PEAKS), plotting_parameter=bad)


def test_mgbt_censoring_flags_the_low_floods_and_matches_an_explicit_threshold():
    low = PEAKS + [120.0, 95.0]
    assert ch.mgbt_test(low) == 2

    s = ch.analysis_data_summary(ch.analysis_data(low, mgbt_low_outliers=True))
    assert s["number_of_low_outliers"] == 2
    assert s["low_outlier_threshold"] == 8900.0
    assert sum(s["is_low_outlier"]) == 2
    assert all(v < 8900.0 for v, f in zip(s["value"], s["is_low_outlier"]) if f)

    # An explicit threshold at the value MGBT chose must censor the same points.
    explicit = ch.analysis_data_summary(ch.analysis_data(low, low_outlier_threshold=8900.0))
    assert explicit["number_of_low_outliers"] == 2
    assert explicit["is_low_outlier"] == s["is_low_outlier"]


def test_a_threshold_censoring_more_than_half_the_record_is_rejected():
    with pytest.raises(Exception, match="50 percent"):
        ch.analysis_data_summary(ch.analysis_data(PEAKS, low_outlier_threshold=20000.0))


def test_threshold_diagnostics_returns_parallel_lists_per_method():
    x = ch.Distribution("GeneralizedPareto", [0, 100, 0.1]).random(300, seed=4242)

    mrl = ch.threshold_diagnostics(x, u_min=0, u_max=200, n_thresholds=8)
    n = len(mrl["threshold"])
    assert 0 < n <= 8
    assert len(mrl["mean_excess"]) == n
    assert len(mrl["exceedance_count"]) == n
    assert mrl["modified_scale"] == []
    assert all(lo < m for lo, m in zip(mrl["lower_ci"], mrl["mean_excess"]))
    assert all(hi > m for hi, m in zip(mrl["upper_ci"], mrl["mean_excess"]))
    counts = mrl["exceedance_count"]
    assert all(b <= a for a, b in zip(counts, counts[1:]))

    st = ch.threshold_diagnostics(x, u_min=0, u_max=200, n_thresholds=8,
                                  method="parameter_stability")
    assert len(st["shape"]) == len(st["threshold"])
    assert st["mean_excess"] == []
    assert all(lo < s for lo, s in zip(st["shape_lower_ci"], st["shape"]))
    # The sample is GPD(0, 100, 0.1), so the fitted shape should sit near 0.1.
    assert abs(st["shape"][0] - 0.1) < 0.15


def test_threshold_diagnostics_validates_its_method():
    with pytest.raises(ValueError, match="unknown method"):
        ch.threshold_diagnostics(PEAKS, 0, 1, method="nope")


def test_repr_shows_the_series_composition():
    assert "15 exact" in repr(ch.analysis_data(PEAKS))
    assert "MGBT" in repr(ch.analysis_data(PEAKS, mgbt_low_outliers=True))
    assert "9000" in repr(ch.analysis_data(PEAKS, low_outlier_threshold=9000.0))


def test_the_frame_round_trips_through_its_json_spec():
    import json

    d = ch.analysis_data(
        exact=PEAKS,
        interval={"index": [15], "lower": [30000.0], "value": [35000.0], "upper": [40000.0]},
    )
    assert json.loads(d.to_json()) == d.spec


# Harricana River annual peaks (Bobee & Ashkar 1991, Table 1.2, page 5); the oracle for
# mann_kendall (0.7757) is ExactDataHypothesisTests.Test_MannKendall, also transcribed into
# fixtures/data/data_frame_facades.json and core/tests/test_data_frame_facades.cpp.
HARRICANA = [
    122, 244, 214, 173, 229, 156, 212, 263, 146, 183, 161, 205, 135, 331, 225,
    174, 98.8, 149, 238, 262, 132, 235, 216, 240, 230, 192, 195, 172, 173, 172,
    153, 142, 317, 161, 201, 204, 194, 164, 183, 161, 167, 179, 185, 117, 192,
    337, 125, 166, 99.1, 202, 230, 158, 262, 154, 164, 182, 164, 183, 171, 250,
    184, 205, 237, 177, 239, 187, 180, 173, 174,
]


def test_analysis_data_hypothesis_test_reproduces_the_harricana_mann_kendall_oracle():
    r = ch.analysis_data_hypothesis_test(HARRICANA, "mann_kendall")
    assert r == {"mann_kendall": pytest.approx(0.7757, abs=1e-3)}


def test_analysis_data_hypothesis_test_accepts_an_analysis_data_frame():
    r_seq = ch.analysis_data_hypothesis_test(HARRICANA, "mann_kendall")
    r_obj = ch.analysis_data_hypothesis_test(ch.analysis_data(HARRICANA), "mann_kendall")
    assert r_seq == r_obj


def test_analysis_data_hypothesis_test_runs_a_two_sample_test_given_a_split_index():
    r = ch.analysis_data_hypothesis_test(HARRICANA, "mann_whitney", split_index=50)
    assert r["mann_whitney"] == pytest.approx(0.5892, abs=1e-2)


def test_analysis_data_hypothesis_test_validates_method_and_the_two_sample_split_index():
    with pytest.raises(ValueError, match="unknown method"):
        ch.analysis_data_hypothesis_test(HARRICANA, "not_a_method")
    with pytest.raises(ValueError, match="requires .split_index."):
        ch.analysis_data_hypothesis_test(HARRICANA, "equal_variance_t")


def test_analysis_data_statistics_returns_the_twenty_named_summary_statistics():
    s = ch.analysis_data_statistics(PEAKS)
    assert len(s["value"]) == 20
    assert list(s["value"].keys()) == [
        "Record Length", "Events Per Index (λ)", "Low Outliers", "Minimum", "Maximum",
        "Mean", "Std Dev", "Skewness", "Kurtosis", "Mean (of log)", "Std Dev (of log)",
        "Skewness (of log)", "Kurtosis (of log)", "1%", "5%", "25%", "50%", "75%", "95%", "99%",
    ]
    assert math.isfinite(s["value"]["Mean"])  # PEAKS has 15 >= 10 exact points


def test_analysis_data_statistics_all_data_uses_the_plotting_position_fit():
    exact_only = ch.analysis_data_statistics(HARRICANA)
    all_data = ch.analysis_data_statistics(HARRICANA, all_data=True)
    assert exact_only["value"]["Record Length"] == 69
    assert exact_only["value"]["Minimum"] == 98.8
    assert all_data["value"]["Minimum"] == 98.8
    # The two methods use different estimators, so their means need not agree even with no
    # censored data in the frame.
    assert exact_only["value"]["Mean"] != pytest.approx(all_data["value"]["Mean"])


def test_analysis_data_statistics_standardized_adds_parallel_columns():
    s = ch.analysis_data_statistics(HARRICANA, standardized=True)
    assert len(s["standardized_value"]) == len(HARRICANA)
    assert len(s["standardized_log10_value"]) == len(HARRICANA)
    assert all(math.isfinite(v) for v in s["standardized_value"])
    assert all(math.isfinite(v) for v in s["standardized_log10_value"])


def test_analysis_data_statistics_reports_nan_for_a_short_record():
    s = ch.analysis_data_statistics(HARRICANA[:9])
    assert all(math.isnan(v) for v in s["value"].values())
